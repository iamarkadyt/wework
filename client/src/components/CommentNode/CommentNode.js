import React from 'react'
import Field from '../Field/Field'
import { connect } from 'react-redux'
import { deleteComment } from '../../state/actions/postsActions'
import { fetchUsersStats } from '../../state/actions/userActions'

const dateFormatOptions = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric'
}

const CommentNode = ({
    _id,
    text,
    date,
    user: {
        _id: authorId,
        name
    },
    postId,
    authedUserId,
    deleteComment,
    fetchUsersStats
}) => {
    const belongsToAuthedUser = authorId === authedUserId
    return (
        <div className="CommentsView-CommentNode-container">
            <p><b>{name}</b> said at {new Date(date)
                .toLocaleDateString('en-US', dateFormatOptions)}:</p>
            <p className="CommentsView-CommentNode-container-Node-body">{text}</p>
            {belongsToAuthedUser
                ? <div>
                    <Field
                        type="linkButton"
                        label="Delete"
                        inline
                        onClick={() =>
                            deleteComment(postId, _id, fetchUsersStats)
                        } />
                </div>
                : null}
        </div>
    )
}

export const _UnconnectedCommentNode = CommentNode
export default connect(state => ({
    authedUserId: state.user.id
}), { deleteComment, fetchUsersStats })(CommentNode)
