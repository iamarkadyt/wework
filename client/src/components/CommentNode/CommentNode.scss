@import 'variables';

.CommentNode-container {
    .CommentNode-container-Node-body {
        font-size: 1.3rem;
    }
    margin-top: 1rem;

    div.Field-container button {
      color: darken($color-font-secondary, 30%);
    }
}
