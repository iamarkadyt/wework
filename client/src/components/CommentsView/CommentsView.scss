.CommentsView-container {
    display: flex;
    justify-content: center;
    .CommentsView-content {
        width: 40rem;
        .CommentsView-comments {
            padding: 1rem;
            margin-top: -2rem;
        }
    }
}

.CommentsView-CommentNode-container {
    .CommentsView-CommentNode-container-Node-body {
        font-size: 1.3rem;
    }
    margin-top: 1rem;
}