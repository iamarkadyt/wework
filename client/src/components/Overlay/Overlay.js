import React from 'react'
import './Overlay.scss'

class Overlay extends React.Component {
  componentWillMount() {
    document.body.style = "overflow: hidden"
  }

  componentWillUnmount() {
    document.body.style = null
  }

  handleSubmit(e) {
    const { 
      onBackdropClick 
    } = this.props

    if (e.target.id === 'Overlay-container-backdrop') {
      onBackdropClick()
    }
  }

  render() {
    const { 
      children, 
      centered
    } = this.props

    return (
      <div id="Overlay-container-backdrop" className="Overlay-container" onClick={this.handleSubmit}>
        <div className={["Overlay-content", centered ? "Overlay-content-centered" : null].join(' ')}>
          {children}
        </div>
      </div>
    )
  }
}

export default Overlay
