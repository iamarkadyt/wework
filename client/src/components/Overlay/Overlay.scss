@import 'variables.scss';
.Overlay-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: calc(#{$navigation_height} + 1rem) 0 2rem 0;
    .content {
        z-index: 1;
        background: white;
        padding: 0.3rem;
        border-radius: .3rem;
        box-shadow: $box_shadow_main;
    }
    .backdrop {
        position: fixed;
        top: 0;
        left: 0;
        width: 100vw;
        height: 100vh;
        background-color: rgba(0, 0, 0, 0.5);
    }
}