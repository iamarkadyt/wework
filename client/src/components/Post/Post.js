import React, { Component } from 'react'
import './Post.scss'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import {
    FaThumbsUp as IcoLike,
    FaComments as IcoComments,
    FaEllipsisV as IcoMore
} from 'react-icons/fa'

import Field from '../Field/Field'
import placeholderImage from '../../images/avatar_placeholder.png'
import { likePost, deleteLike, deletePost } from '../../state/actions/postsActions'
import {
    unfollowAPerson,
    fetchDiscoverContent,
    fetchUsersStats
} from '../../state/actions/userActions'

class Post extends Component {
    state = {
        showMenu: false
    }

    dismissMenu = e => {
        if (e.target.className !== "Post__button--menu")
            this.setState({ showMenu: false })
    }

    componentWillMount() {
        window.addEventListener('click', this.dismissMenu)
    }

    componentWillUnmount() {
        window.removeEventListener('click', this.dismissMenu)
    }

    render() {
        const {
            _id,
            user: {
                _id: authorId,
                name,
                avatar
            },
            text,
            comments,
            likes,
            date,
            authedUser,
            likePost,
            deleteLike,
            deletePost,
            unfollowAPerson,
            fetchDiscoverContent,
            fetchUsersStats,
            history,
            match,
            flat,
            nocomments
        } = this.props
        // likes.user field does not get populated and directly contains the _id
        const likedByAuthedUser = !!likes.find(item => item.user === authedUser.id)

        const belongsToAuthedUser = authorId === authedUser.id
        const baseUrl = match.url || ''
        const dateFormatOptions = {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
            hour: 'numeric',
            minute: 'numeric',
            second: 'numeric'
        }
        const flatStyle = {
            borderRadius: 'unset',
            boxShadow: 'unset',
            marginTop: 'unset',
        }

        return <div className="Post__container" style={flat ? flatStyle : null}>
            <div className={["Post__menu",
                this.state.showMenu
                    ? "Post__menu--shown"
                    : ""].join(' ')}>
                {belongsToAuthedUser
                    ? <button
                        className="Post__menu-item"
                        onClick={() => 
                            deletePost(_id, () => {
                                fetchUsersStats()
                            })}>
                        Delete</button>
                    : <button
                        className="Post__menu-item"
                        onClick={() =>
                            unfollowAPerson(authorId, () => {
                                fetchUsersStats()
                                fetchDiscoverContent(5)
                            })
                        }>
                        Unfollow user</button>}
            </div>
            <div className="Post__header">
                <img className="Post__avatar" src={avatar || placeholderImage} alt='' />
                <button
                    className="Post__button--menu"
                    onClick={() => this.setState(
                        prevState => ({ showMenu: !prevState.showMenu })
                    )}>
                    <IcoMore />
                </button>
                <div className="Post__name">
                    <Field
                        type="linkButton"
                        label={name}
                        inline
                        style={{ fontSize: '1.25rem' }}
                        onClick={() => history.push(
                            belongsToAuthedUser
                                ? '/profile'
                                : `/profile/id/${authorId}`
                        )} />
                </div>
                <p className="Post__date">{new Date(date).toLocaleDateString('en-US', dateFormatOptions)}</p>
            </div>
            <p className="Post__body">{text}</p>
            <div className="Post__stats">
                <span>{likes.length} likes &nbsp;â€¢&nbsp; {comments.length} comments</span>
            </div>
            <div className="Post__buttons">
                <Field
                    type="linkButton"
                    inline
                    style={ likedByAuthedUser ? { color: 'cornflowerblue' } : null }
                    onClick={() => {
                        likedByAuthedUser
                            ? deleteLike(_id, fetchUsersStats)
                            : likePost(_id, fetchUsersStats)
                    }}>
                    <span className="Post__buttons-icon"><IcoLike /></span>
                    &nbsp;Like
            </Field>
                {nocomments
                    ? null
                    : <Field
                        type="linkButton"
                        inline
                        style={{ marginLeft: '1rem' }}
                        onClick={() => history.push(`${baseUrl}/view-comments/${_id}`)}>
                        <span className="Post__buttons-icon"><IcoComments /></span>
                        &nbsp;Comment
            </Field>}
            </div>
        </div>
    }
}

export const _UnconnectedPost = Post;
export default withRouter(connect(state => ({
    authedUser: state.user
}), {
        likePost,
        deleteLike,
        deletePost,
        unfollowAPerson,
        fetchDiscoverContent,
        fetchUsersStats
    })(Post))
