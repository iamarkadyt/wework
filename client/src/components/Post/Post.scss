@import 'variables.scss';
.Post__container {
    position: relative;
    border-radius: .3rem;
    overflow: hidden;
    box-shadow: $box_shadow_main;
    margin-top: 1rem;
}

.Post__header {
    padding: 1rem 1rem 0 1rem;
}

.Post__avatar {
    float: left;
    width: 3rem;
    border-radius: 3rem;
    margin-right: 1rem;
}

.Post__name {
    font-weight: bolder;
    font-size: 1.2rem;
}

.Post__body {
    padding: 1rem;
    font-size: 1.3rem;
}

.Post__stats {
    padding: 0 1rem 1rem 1rem;
}

.Post__buttons {
    padding: 0 1rem 1rem 1rem;
}

.Post__buttons-icon {
    font-size: .8rem;
}

.Post__button--menu {
    float: right;
    background: none;
    border: none;
    font-size: 1rem;
    color: rgba(0, 0, 0, 0.5);
    padding: .5rem;
    cursor: pointer;
    &:hover {
        color: rgba(0, 0, 0, 0.8)
    }
    &:hover .Post__menu {
        position: relative;
    }
    svg {
        pointer-events: none;
    }
}

.Post__menu {
    opacity: 0;
    pointer-events: none;
    transition: .1s all ease-out;
    position: absolute;
    top: 2rem;
    right: 2rem;
    background-color: white;
    border-radius: .3rem;
    overflow: hidden;
    box-shadow: $box_shadow_main;
}

.Post__menu--shown {
    opacity: 1;
    pointer-events: unset;
}

.Post__menu-item {
    display: block;
    width: 100%;
    text-align: left;
    background: none;
    border: none;
    cursor: pointer;
    font-size: 1.3rem;
    padding: .5rem 1rem;
    &:hover {
        background-color: darken(white, 10%);
    }
}