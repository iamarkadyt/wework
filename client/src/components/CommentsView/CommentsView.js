import React, { Component } from 'react'
import './CommentsView.scss'
import { connect } from 'react-redux'
import { func, object } from 'prop-types'

import Post from '../Post/Post'
import Reply from '../Reply/Reply'
import Overlay from '../Overlay/Overlay'
import { addComment } from '../../state/actions/postsActions'
import CommentNode from '../CommentNode/CommentNode'
import { postType } from '../../types/index'

/**
 * Component's contract
 *
 * What does my component render?
 * - Post component with flat & nocomments modflags,
 * - Reply component
 * - Array of CommentNodes
 * - All of the above is rendered in Overlay.
 *
 * Ok, but what does it render differently in different circumstances? (Props)
 * > Sure, look:
 * - If there is no post data upon the update, comment navigates back and returns false at shldCompUpd()
 * - If there is post data, component proceedes with the regular render
 * - If there are no post.comments component renders message instead of comments (lol).
 * - If there are post.comments, component maps the array to CommentNodes
 *
 * What about functions passed as props?
 * - addComent() is expected to be called upon the submisson 
 * - and is expected to be called with correct arguments: postId, commentBody, serviceCallback
 *
 * Interaction!
 * - No interactable parts.
 *
 * State?
 * - No
 *
 * List of constraints:
 * - If there is no post data upon the update, comment navigates back and returns false at shldCompUpd()
 * - If there is post data, component proceedes with the regular render
 * - If there are no post.comments component renders message instead of comments (lol).
 * - If there are post.comments, component maps the array to CommentNodes
 * - addComent() is expected to be called upon the submisson 
 * - and is expected to be called with correct arguments: postId, commentBody, serviceCallback
 */

class CommentsView extends Component {
    handleDismiss = () => {
        const { history } = this.props
        history.goBack()
    }

    shouldComponentUpdate(nextProps) {
        if (!nextProps.post) {
            this.handleDismiss()
            return false
        }
        return true
    }

    render() {
        const {
            post,
            post: {
                _id,
                comments
            },
            addComment
        } = this.props

        return (
            <Overlay onBackdropClick={() => this.handleDismiss()}>
                <div className="CommentsView-container">
                    <div className="CommentsView-content">
                        <Post {...post} flat nocomments />
                        <Reply
                            rows="3"
                            flat
                            onSubmit={(data, callback) => addComment(_id, data, callback)} />
                        <div className="CommentsView-comments">
                            <h2>Comments:</h2>
                            {comments.length > 0
                                ? comments.map(item => (
                                    <CommentNode
                                        key={item._id}
                                        postId={_id}
                                        {...item} />
                                ))
                                : <p>Be the first one to leave a comment!</p>}
                        </div>
                    </div>
                </div>
            </Overlay>
        )
    }
}

CommentsView.propTypes = {
  post: postType,
  addComment: func.isRequired,
  history: object.isRequired
}

export const _UnconnectedCommentsView = CommentsView
export default connect(null, { addComment })(CommentsView)
