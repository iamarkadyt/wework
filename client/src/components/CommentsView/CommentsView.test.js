import React from 'react'
import { _UnconnectedCommentsView as CommentsView } from './CommentsView'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'
import { mockPost } from '../../mocks/posts'
import Reply from '../Reply/Reply'
import Overlay from '../Overlay/Overlay'
import Post from '../Post/Post'
import CommentNode from '../CommentNode/CommentNode'

describe('CommentsView', () => {
  let props
  let mountedCommentsView

  const getMockProps = () => {
    return cloneDeep({
      history: {
        goBack: jest.fn()
      },
      post: mockPost,
      addComment: jest.fn()
    })
  }

  const commentsView = () => {
    if (!mountedCommentsView) {
      mountedCommentsView = shallow(<CommentsView {...props} />)
    }
    return mountedCommentsView
  }

  beforeEach(() => {
    props = getMockProps()
    mountedCommentsView = undefined
  })

  it('handleDismiss() function must operate as expected', () => {
    commentsView().instance().handleDismiss()
    expect(props.history.goBack).toHaveBeenCalled()
  })

  describe('shouldComponentUpdate', () => {
    it('blocks an update and navigates back if post object wasn\'t found in nextProps', () => {
      commentsView().setProps({ post: undefined })
      expect(props.history.goBack).toHaveBeenCalled()
    })

    it('doesn\'t block an update if post object was found in nextProps', () => {
      expect(commentsView().instance().shouldComponentUpdate(props, null)).toBe(true)
    })
  })

  describe('comments', () => {
    it('are replaced with a message when post has none', () => {
      props.post.comments = []
      expect(commentsView().find(".CommentsView-comments").childAt(1).name()).toBe('p')
    })

    it('are rendered if post has them', () => {
      expect(commentsView().find(".CommentsView-comments").find(CommentNode).length).toBe(2)
    })
  })

  describe('addComment()', () => {
    it('must be called upon the comment submission', () => {
      commentsView().find(Reply).simulate('submit')
      expect(props.addComment).toHaveBeenCalled()
    })

    it('must be called with correct parms', () => {
      const postId = 'some_post_id123'
      props.post._id = postId
      commentsView().find(Reply).simulate('submit', 'data', () => {})
      expect(props.addComment).toHaveBeenCalledWith(postId, 'data', expect.any(Function))
    })
  })

  describe('props passing', () => {
    it('Post component receives correct props', () => {
      const expectedProps = { ...props.post, flat: true, nocomments: true }
      expect(commentsView().find(Post).props()).toStrictEqual(expectedProps)
    })

    it('Reply component receives correct props', () => {
      // submit handler is tested elsewhere
      const expectedProps = { rows: "3", flat: true, onSubmit: expect.any(Function) }
      expect(commentsView().find(Reply).props()).toStrictEqual(expectedProps)
    })

    it('CommentNode component receives correct props', () => {
      // only have to verify a single case to prove that all uses in this component
      // would work as expected; do not include key prop in expectedProps.
      const comment = props.post.comments[0]
      const expectedProps = { postId: props.post._id, comment }
      expect(commentsView().find(CommentNode).first().props()).toStrictEqual(expectedProps)
    })

    it('Overlay component receives correct props', () => {
      const expectedProps = { 
        onBackdropClick: expect.any(Function),
        children: expect.anything()
      }
      expect(commentsView().find(Overlay).props()).toStrictEqual(expectedProps)
    })

    it('Overlay\'s handleDismiss() is called on backdrop click', () => {
      commentsView().find(Overlay).simulate('backdropClick')
      expect(props.history.goBack).toHaveBeenCalled()
    })
  })

  it('must pass a snapshot test', () => {
    // also verifies for the following constraints:
    // - if there are comments, comp should render them
    // - if there is post data, comp should render it
    // warning: enzyme-to-json is used
    expect(commentsView()).toMatchSnapshot()
  })
})

/**
 * Conclusions:
 * 
 * Not only the effects of passed props should be tested,
 * but also that internal components receive right props.
 */
