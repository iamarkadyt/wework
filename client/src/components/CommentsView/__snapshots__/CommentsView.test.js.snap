// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsView must pass a snapshot test 1`] = `
<Overlay
  onBackdropClick={[Function]}
>
  <div
    className="CommentsView-container"
  >
    <div
      className="CommentsView-content"
    >
      <withRouter(Connect(Post))
        __v={0}
        _id="5b9f4728bc40273febf52706"
        comments={
          Array [
            Object {
              "_id": "5ba1dd77809c892f580240ef",
              "date": "2018-09-19T05:24:07.399Z",
              "text": "Hahahahahahahaaa!",
              "user": Object {
                "_id": "5b75f023f1957052e52106f5",
                "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
                "name": "John Grapard",
              },
            },
            Object {
              "_id": "5ba1de21f0b9f73054396abf",
              "date": "2018-09-19T05:26:57.632Z",
              "text": "Nuka proverim esli oshibka ostalas...",
              "user": Object {
                "_id": "5b75f023f1957052e52106f5",
                "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
                "name": "John Grapard",
              },
            },
          ]
        }
        date="2018-09-17T06:18:16.992Z"
        flat={true}
        likes={
          Array [
            Object {
              "_id": "5ba1dd61809c892f580240ed",
              "user": "5b75f023f1957052e52106f5",
            },
            Object {
              "_id": "5ba80533058247301e362bf9",
              "user": "5ba1acfc13b542138b29af68",
            },
            Object {
              "_id": "5ba847153c823e00161763b5",
              "user": "5ba83562ed730446b54def34",
            },
          ]
        }
        nocomments={true}
        text="Tas nisl commlor sit amet, consec!"
        user={
          Object {
            "_id": "5b9eedd621194c102c5280e8",
            "avatar": "//www.gravatar.com/avatar/e13743a7f1db7f4246badd6fd6ff54ff?s=200&r=pg&d=mm",
            "name": "John Doe",
          }
        }
      />
      <Connect(Reply)
        flat={true}
        onSubmit={[Function]}
        rows="3"
      />
      <div
        className="CommentsView-comments"
      >
        <h2>
          Comments:
        </h2>
        <Connect(CommentNode)
          _id="5ba1dd77809c892f580240ef"
          date="2018-09-19T05:24:07.399Z"
          key="5ba1dd77809c892f580240ef"
          postId="5b9f4728bc40273febf52706"
          text="Hahahahahahahaaa!"
          user={
            Object {
              "_id": "5b75f023f1957052e52106f5",
              "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
              "name": "John Grapard",
            }
          }
        />
        <Connect(CommentNode)
          _id="5ba1de21f0b9f73054396abf"
          date="2018-09-19T05:26:57.632Z"
          key="5ba1de21f0b9f73054396abf"
          postId="5b9f4728bc40273febf52706"
          text="Nuka proverim esli oshibka ostalas..."
          user={
            Object {
              "_id": "5b75f023f1957052e52106f5",
              "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
              "name": "John Grapard",
            }
          }
        />
      </div>
    </div>
  </div>
</Overlay>
`;
